airbyte-workers/src/main/java/io/airbyte/workers/process/AsyncOrchestratorPodProcess.java

```diff
@@ -0,0 +1,282 @@
+  private final KubernetesClient kubernetesClient;
+  private final AtomicReference<Optional<Integer>> cachedExitValue;
+
+  public AsyncOrchestratorPodProcess(
+                                     final KubePodInfo kubePodInfo,
+                                     final DocumentStoreClient documentStoreClient,
+                                     final KubernetesClient kubernetesClient) {
+    this.kubePodInfo = kubePodInfo;
+    this.documentStoreClient = documentStoreClient;
+    this.kubernetesClient = kubernetesClient;
+    this.cachedExitValue = new AtomicReference<>(Optional.empty());
+  }
+
+  public Optional<String> getOutput() {
+    return getDocument(AsyncKubePodStatus.SUCCEEDED.name());
+  }
+
+  private int computeExitValue() {
+    final AsyncKubePodStatus docStoreStatus = getDocStoreStatus();
+
+    // trust the doc store if it's in a terminal state
+    if (docStoreStatus.equals(AsyncKubePodStatus.FAILED)) {
+      return 1;
+    } else if (docStoreStatus.equals(AsyncKubePodStatus.SUCCEEDED)) {
+      return 0;
+    }
+
+    final Pod pod = kubernetesClient.pods()
+        .inNamespace(getInfo().namespace())
+        .withName(getInfo().name())
+        .get();
+
+    // Since the pod creation blocks until the pod is created the first time,
+    // if the pod no longer exists (and we don't have a success/fail document)
+    // we must be in a state where it completed and therefore failed
+    if (pod == null) {
+      return 1;
+    }
+
+    // If the pod does exist, it may be in a terminal (error or completed) state.
+    final boolean isTerminal = KubePodProcess.isTerminal(pod);
+
+    if (isTerminal) {
+      // In case the doc store was updated in between when we pulled it and when
+      // we read the status from the Kubernetes API, we need to check the doc store again.
+      final AsyncKubePodStatus secondDocStoreStatus = getDocStoreStatus();
+      if (secondDocStoreStatus.equals(AsyncKubePodStatus.FAILED)) {
+        return 1;
+      } else if (secondDocStoreStatus.equals(AsyncKubePodStatus.SUCCEEDED)) {
+        return 0;
+      } else {
+        // otherwise, the actual pod is terminal when the doc store says it shouldn't be.
+        return 1;
+      }
+    }
+
+    // Otherwise, throw an exception because this is still running.
+    switch (docStoreStatus) {
+      case NOT_STARTED -> throw new IllegalThreadStateException("Pod hasn't started yet.");
+      case INITIALIZING -> throw new IllegalThreadStateException("Pod is initializing.");
+      default -> throw new IllegalThreadStateException("Pod is running.");
+    }
+  }
+
+  @Override
+  public int exitValue() {
+    final var optionalCached = cachedExitValue.get();
+
+    if (optionalCached.isPresent()) {
+      return optionalCached.get();
+    } else {
+      final var exitValue = computeExitValue();
+      cachedExitValue.set(Optional.of(exitValue));
+      return exitValue;
+    }
+  }
+
+  @Override
+  public void destroy() {
```

Comment:
Should we delete the doc store here

Reply:
I think there's no downside to maintaining a history on cloud storage? Maybe we'll need to add this to job sweeping eventually.

Reply:
Agree we can create an issue to remind ourselves and punt this till later.